src/merger/merger.c

```diff
@@ -597,6 +597,19 @@ luaL_merge_source_buffer_next(struct merge_source *base,
 	*rpos = (char *)tuple_end;
 	if (format == NULL)
 		format = box_tuple_format_default();
+	/*
+	 * If we encounter an MP_TUPLE, skip the extension header and the tuple
+	 * format identifier.
+	 */
+	if (mp_typeof(*tuple_beg) == MP_EXT) {
+		uint32_t len;
+		int8_t type;
+		mp_decode_ext(&tuple_beg, &type, &len);
+		assert(tuple_beg + len == tuple_end);
+		assert(mp_typeof(*tuple_beg) == MP_UINT);
+		/* Skip the tuple format identifier. */
+		mp_decode_uint(&tuple_beg);
+	}
```

Comment:
Shouldn't we perform this logic only if the the extension is of type MP_TUPLE?

I know that it wouldn't change the behavior much due to https://github.com/tarantool/tarantool/issues/5048. However, it at least wouldn't make the problem deeper.

Reply:
How do we know we know what is the code of `MP_TUPLE`? Should I hardcode it?

Reply:
> Shouldn't we perform this logic only if the the extension is of type MP_TUPLE?

I expected that only the `MP_TUPLE` extension can lead to this branch (in the embedded version I have an assertion on this). Not clear what to do if we get a different extension type.

Reply:
> How do we know we know what is the code of `MP_TUPLE`? Should I hardcode it?

Sure. The extension codes are part of the public binary protocol. See https://www.tarantool.io/en/doc/latest/dev_guide/internals/msgpack_extensions/

> I expected that only the `MP_TUPLE` extension can lead to this branch (in the embedded version I have an assertion on this). Not clear what to do if we get a different extension type.

That's so if the input buffer is well-formed. However, we shouldn't crash if it is ill-formed. We should report an error instead.

Reply:
> I added an `mp_extension_type` enum for the `MP_TUPLE` code, and now I set a diagnostic and return an error when the extension type is wrong. 

It looks correct for me. Thanks!